chain/client/src/stateless_validation/processing_tracker.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+
+use once_cell::sync::OnceCell;
+
+/// `ProcessingDoneTracker` can be used in conjuction with a `ProcessingDoneWaiter`
+/// to wait until some processing is finished. `ProcessingDoneTracker` should be
+/// kept alive as long as the processing is ongoing, then once it's dropped,
+/// the paired `ProcessingDoneWaiter` will be notified that the processing has finished.\
+/// Does NOT implement `Clone`, if you want to clone it, wrap it in an `Arc`.
+pub struct ProcessingDoneTracker(Arc<OnceCell<()>>);
+
+impl ProcessingDoneTracker {
+    /// Create a new `ProcessingDoneTracker`
+    pub fn new() -> ProcessingDoneTracker {
+        ProcessingDoneTracker(Arc::new(OnceCell::new()))
+    }
+
+    /// Create a `ProcessingDoneWaiter` paired to this `ProcessingDoneTracker`.\
+    /// When this `ProcessingDoneTracker` is dropped, the paired waiter will be notified.
+    pub fn make_waiter(&self) -> ProcessingDoneWaiter {
+        ProcessingDoneWaiter(self.0.clone())
+    }
+}
+
+impl Drop for ProcessingDoneTracker {
+    fn drop(&mut self) {
+        // Set a value to notify waiters that the processing has finished.
+        //
+        // OnceCell::set() returns an Err when the cell already has a value.
+        // This should be impossible, as there's only one ProcessingDoneTracker
+        // that will set this value when dropped, so it's safe to unwrap() it.
+        self.0.set(()).unwrap();
+    }
+}
+
+/// `ProcessingDoneWaiter` is used to wait until the processing has finished.\
+/// The `wait()` method will block until the paired `ProcessingDoneTracker` is dropped.\
+/// A new instance of `ProcessingDoneWaiter` can be created by calling `ProcessingDoneTracker::make_waiter`.
+#[derive(Clone)]
+pub struct ProcessingDoneWaiter(Arc<OnceCell<()>>);
```

Comment:
```suggestion
pub struct ProcessingDoneWaiter(ProcessingDoneTracker);
```
With that, connection between waiter and tracker is more clear, and we could use common-used `From` to create waiter from tracker.

Reply:
But then when a waiter is dropped it would also drop the tracker, that would break the functionality.

Reply:
Oh. Now I was going to propose `Arc<ProcessingDoneTracker>` and only then I understood how exactly it works 👍